ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/debug/DebuggerStateManager.java

```diff
@@ -0,0 +1,161 @@
+    private final LocalStorage                      localStorage;
+    private final DtoFactory                        dtoFactory;
+    private final String                            storageUniqueKey;
+    private final String                            workspaceId;
+
+    @Inject
+    public DebuggerStateManager(AppContext appContext,
+                                WorkspaceServiceClient workspaceServiceClient,
+                                EventBus eventBus,
+                                LocalStorageProvider localStorageProvider,
+                                DtoFactory dtoFactory) {
+        this.eventBus = eventBus;
+        this.localStorage = localStorageProvider.get();
+        this.dtoFactory = dtoFactory;
+        this.workspaceId = appContext.getWorkspaceId();
+        this.storageUniqueKey = LOCAL_STORAGE_DEBUGGER_STATES_KEY_PREFIX + workspaceId;
+        cleanup(workspaceServiceClient);
+        addHandlers();
+    }
+
+    /**
+     * Returns debugger state for given debugger type in scope of current workspace.
+     * 
+     * @param debuggerType
+     * @return debugger state
+     */
+    public DebugSessionStateDto getDebuggerState(String debuggerType) {
+        List<DebugSessionStateDto> debugSessionStates = readStateData();
+        return findStoredState(debugSessionStates, debuggerType);
+    }
+
+    /**
+     * Sets debugger state for given debugger type in scope of current workspace.
+     * 
+     * @param debugSessionState
+     */
+    public void setDebuggerState(DebugSessionStateDto debugSessionState) {
+        List<DebugSessionStateDto> debugSessionStates = readStateData();
+        DebugSessionState matchingSessionState =
+                                               findStoredState(debugSessionStates, debugSessionState.getDebuggerType());
+        debugSessionStates.remove(matchingSessionState);
+        debugSessionStates.add(debugSessionState);
+        writeStateData(debugSessionStates);
+    }
+
+    /**
+     * Removes debugger state for given debugger type in scope of current workspace.
+     * 
+     * @param debuggerType
+     */
+    public void removeDebuggerState(String debuggerType) {
+        List<DebugSessionStateDto> debugSessionStates = readStateData();
+        DebugSessionState matchingSessionState = findStoredState(debugSessionStates, debuggerType);
+        debugSessionStates.remove(matchingSessionState);
+        writeStateData(debugSessionStates);
+    }
+
+    private void addHandlers() {
+        eventBus.addHandler(WorkspaceStoppedEvent.TYPE, new WorkspaceStoppedEvent.Handler() {
+            @Override
+            public void onWorkspaceStopped(WorkspaceStoppedEvent event) {
+                if (workspaceId.equals(event.getWorkspace().getId())) {
+                    // Wipe out storage data key on workspace stop
+                    writeStateData(null);
+                }
+            }
+        });
+    }
+
+    private List<DebugSessionStateDto> readStateData() {

```

Comment:
I am not sure what is the problem here... As states data is stored as a list via what is simply achievable with `DtoFactory` (can not see anything for storing maps there) I would need to create a map basing on this list and then get the values back from it via `Map.values()` to put it back to `writeStateData()?` Just to get rid of `findStoredState?`